File: mcp/sonarqube-mcp/src/logging.ts
/**
 * Logging for sonarqube-mcp.
 *
 * The server speaks JSON-RPC to Claude Code over stdio, so **stdout must stay
 * pristine** — a stray `console.log` would corrupt the protocol stream. Every
 * log line here is written to stderr.
 */

export const LOG_LEVELS = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'] as const;

export type LogLevel = (typeof LOG_LEVELS)[number];

const LEVEL_RANK: Record<LogLevel, number> = {
  DEBUG: 10,
  INFO: 20,
  WARNING: 30,
  ERROR: 40,
  CRITICAL: 50,
};

let currentLevel: LogLevel = 'INFO';

/** Set the global log level. Safe to call repeatedly. */
export function configureLogging(level: LogLevel = 'INFO'): void {
  currentLevel = level;
}

function emit(level: LogLevel, name: string, message: string): void {
  if (LEVEL_RANK[level] < LEVEL_RANK[currentLevel]) return;
  const timestamp = new Date().toISOString();
  // stderr only — see the module docstring.
  process.stderr.write(`${timestamp} | ${level.padEnd(8)} | ${name} | ${message}\n`);
}

export interface Logger {
  debug(message: string): void;
  info(message: string): void;
  warning(message: string): void;
  error(message: string): void;
  critical(message: string): void;
}

/** Return a module-scoped logger. */
export function getLogger(name: string): Logger {
  return {
    debug: (message) => emit('DEBUG', name, message),
    info: (message) => emit('INFO', name, message),
    warning: (message) => emit('WARNING', name, message),
    error: (message) => emit('ERROR', name, message),
    critical: (message) => emit('CRITICAL', name, message),
  };
}

File: mcp/sonarqube-mcp/src/sonarqube_mcp/logging.py
"""Logging configuration for sonarqube-mcp.

The server communicates with Claude Code over stdio using JSON-RPC, so
**stdout must stay pristine** — any stray print() or log line written to
stdout would corrupt the protocol stream. All logging in this package is
therefore configured to write to stderr only.
"""

from __future__ import annotations

import logging
import sys

_CONFIGURED = False

_LOG_FORMAT = "%(asctime)s | %(levelname)-8s | %(name)s | %(message)s"
_DATE_FORMAT = "%Y-%m-%dT%H:%M:%S%z"


def configure_logging(level: str = "INFO") -> None:
    """Configure the root logger once, idempotently.

    Safe to call multiple times (e.g. from tests and from ``server.main``);
    subsequent calls only adjust the log level, they never attach duplicate
    handlers.
    """
    global _CONFIGURED

    root = logging.getLogger()
    normalized_level = getattr(logging, level.upper(), logging.INFO)

    if not _CONFIGURED:
        handler = logging.StreamHandler(stream=sys.stderr)
        handler.setFormatter(logging.Formatter(_LOG_FORMAT, datefmt=_DATE_FORMAT))
        root.addHandler(handler)
        _CONFIGURED = True

        # Third-party libraries are noisy at INFO/DEBUG; keep them quiet
        # unless the operator explicitly asks for DEBUG output.
        for noisy in ("httpx", "httpcore", "hpack"):
            logging.getLogger(noisy).setLevel(logging.WARNING)

    root.setLevel(normalized_level)


def get_logger(name: str) -> logging.Logger:
    """Return a module-scoped logger. Call :func:`configure_logging` first."""
    return logging.getLogger(name)
